File: backend-python/tests/integration/db_fixture.py
from __future__ import annotations

import time
import uuid

import aiomysql
from pymysql.constants import CLIENT

from app.config import Config


class DbTestFixture:
    """実DB(docker-compose上のMySQL)結合テスト共通のfixture。
    テストごとに一意なemail/keycloak_subでユーザー・ラベルを作り、共有の開発用DBを汚さない
    (backend-java/backend-kotlin/backend-cpp/backend-cのDbTestFixture/db_fixtureと同じ設計。
    過去にbackend-rustで固定fixture行の共有が原因のテスト間干渉が見つかった経緯があるため、
    必ずテストごとに一意な行を作る)
    """

    def __init__(self, pool: aiomysql.Pool) -> None:
        self.pool = pool

    @staticmethod
    def unique_suffix() -> str:
        return f"{int(time.time() * 1_000_000)}-{uuid.uuid4().hex[:8]}"

    async def create_user(self, unique_suffix: str) -> int:
        async with self.pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    "INSERT INTO users (email, name, role, created_at, updated_at) VALUES (%s, %s, 1, NOW(), NOW())",
                    (f"backend-python-test-{unique_suffix}@example.com", f"backend-python-test-{unique_suffix}"),
                )
                return cur.lastrowid

    async def create_user_with_keycloak_sub(self, unique_suffix: str, keycloak_sub: str) -> int:
        user_id = await self.create_user(unique_suffix)
        async with self.pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    "INSERT INTO user_keycloaks (user_id, keycloak_sub, created_at, updated_at) "
                    "VALUES (%s, %s, NOW(), NOW())",
                    (user_id, keycloak_sub),
                )
        return user_id

    async def create_label(self, unique_suffix: str) -> int:
        async with self.pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    "INSERT INTO labels (name, created_at, updated_at) VALUES (%s, NOW(), NOW())",
                    (f"backend-python-test-label-{unique_suffix}",),
                )
                return cur.lastrowid

    async def cleanup_user(self, user_id: int) -> None:
        """ベストエフォートの後始末(backend-java/backend-kotlin/backend-c/backend-cppと同じ方針)"""
        try:
            async with self.pool.acquire() as conn:
                async with conn.cursor() as cur:
                    await cur.execute(
                        "DELETE FROM task_labels WHERE task_id IN (SELECT id FROM tasks WHERE user_id = %s)",
                        (user_id,),
                    )
                    await cur.execute("DELETE FROM tasks WHERE user_id = %s", (user_id,))
                    await cur.execute("DELETE FROM user_keycloaks WHERE user_id = %s", (user_id,))
                    await cur.execute("DELETE FROM users WHERE id = %s", (user_id,))
        except Exception:
            pass

    async def cleanup_label(self, label_id: int) -> None:
        try:
            async with self.pool.acquire() as conn:
                async with conn.cursor() as cur:
                    await cur.execute("DELETE FROM labels WHERE id = %s", (label_id,))
        except Exception:
            pass

    async def count_task_labels(self, task_id: int) -> int:
        async with self.pool.acquire() as conn:
            async with conn.cursor() as cur:
                await cur.execute("SELECT COUNT(*) FROM task_labels WHERE task_id = %s", (task_id,))
                row = await cur.fetchone()
                return row[0]


async def build_test_pool() -> aiomysql.Pool:
    config = Config.from_env()
    return await aiomysql.create_pool(
        host=config.db_host,
        port=config.db_port,
        user=config.db_user,
        password=config.db_password,
        db=config.db_schema,
        minsize=1,
        maxsize=5,
        autocommit=True,
        # app/main.pyのbuild_pool()と同じ理由(UPDATE文のrowcountを「マッチ行数」にするため)
        client_flag=CLIENT.FOUND_ROWS,
    )

File: backend-cpp/tests/db_fixture.cpp
#include "db_fixture.hpp"

#include <mysql/mysql.h>

#include <atomic>
#include <chrono>
#include <cstdio>
#include <stdexcept>

namespace backend_cpp::testutil {

namespace {

std::string UniqueSuffix() {
  static std::atomic<int64_t> counter{0};
  auto now_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
                    std::chrono::system_clock::now().time_since_epoch())
                    .count();
  return std::to_string(now_ns) + "-" + std::to_string(counter.fetch_add(1));
}

void MustExec(MYSQL* conn, const std::string& sql) {
  if (mysql_query(conn, sql.c_str()) != 0) {
    throw std::runtime_error(std::string("query failed: ") + mysql_error(conn) + " (" + sql + ")");
  }
}

// エスケープ済みのSQLリテラル文字列を組み立てる(このテストヘルパー限定の簡易版、
// プレースホルダのprepared statementを使うほどの複雑さが無いため生SQLで組み立てる)
std::string Escape(MYSQL* conn, const std::string& s) {
  std::string out(s.size() * 2 + 1, '\0');
  unsigned long len = mysql_real_escape_string(conn, out.data(), s.c_str(), s.size());
  out.resize(len);
  return out;
}

}  // namespace

TestUser::TestUser(db::ConnectionPool& pool, std::optional<std::string> keycloak_sub) : pool_(pool) {
  auto lease = pool_.Acquire();
  MYSQL* conn = lease.get();
  std::string suffix = UniqueSuffix();
  std::string email = "backend-cpp-test-" + suffix + "@example.com";
  std::string name = "backend-cpp-test-" + suffix;
  MustExec(conn, "INSERT INTO users (email, name, role, created_at, updated_at) VALUES ('" +
                     Escape(conn, email) + "', '" + Escape(conn, name) + "', 1, NOW(), NOW())");
  id_ = static_cast<int64_t>(mysql_insert_id(conn));

  if (keycloak_sub.has_value()) {
    MustExec(conn, "INSERT INTO user_keycloaks (user_id, keycloak_sub, created_at, updated_at) "
                   "VALUES (" +
                       std::to_string(id_) + ", '" + Escape(conn, *keycloak_sub) +
                       "', NOW(), NOW())");
  }
}

TestUser::~TestUser() {
  if (id_ == 0) return;
  try {
    auto lease = pool_.Acquire();
    MYSQL* conn = lease.get();
    MustExec(conn, "DELETE FROM task_labels WHERE task_id IN (SELECT id FROM tasks WHERE user_id = " +
                       std::to_string(id_) + ")");
    MustExec(conn, "DELETE FROM tasks WHERE user_id = " + std::to_string(id_));
    MustExec(conn, "DELETE FROM user_keycloaks WHERE user_id = " + std::to_string(id_));
    MustExec(conn, "DELETE FROM users WHERE id = " + std::to_string(id_));
  } catch (...) {
    // 後片付けの失敗でテスト自体を落とさない(ベストエフォート)
  }
}

TestLabel::TestLabel(db::ConnectionPool& pool) : pool_(pool) {
  auto lease = pool_.Acquire();
  MYSQL* conn = lease.get();
  std::string name = "backend-cpp-test-label-" + UniqueSuffix();
  MustExec(conn, "INSERT INTO labels (name, created_at, updated_at) VALUES ('" +
                     Escape(conn, name) + "', NOW(), NOW())");
  id_ = static_cast<int64_t>(mysql_insert_id(conn));
}

TestLabel::~TestLabel() {
  if (id_ == 0) return;
  try {
    auto lease = pool_.Acquire();
    MYSQL* conn = lease.get();
    MustExec(conn, "DELETE FROM task_labels WHERE label_id = " + std::to_string(id_));
    MustExec(conn, "DELETE FROM labels WHERE id = " + std::to_string(id_));
  } catch (...) {
  }
}

}  // namespace backend_cpp::testutil
